Rebase 20160411

deads2k opened the pull request:
Before rebasing on master, all code compiled, and all tests passed.

 - [ ] enable third party resources.  @liggitt wants API changes first and we probably need a full cached discovery client to avoid tons of round trips.
 - [ ] migrate to the new deep copy generator.  The old one is still work for us, but it isn't being improved.  The new deep copy generator had different assumptions about packages.  It wasn't a drop-in.
 - [ ] enable recursive directory options in our cli where appropriate.
 - [ ] @liggitt wants to have a single authorization attributes interface for both kube and origin.  That means changing the origin sig in a non-trivial way.

@pweil- Ready for review for whichever victim you choose.

deads2k commented:
[test]

deads2k commented:
@liggitt @ncdc @dgoodwin @sosiouxme @bparees 

Each of you has a commit with your name/ID in all caps.  Please look at the those commits and review or provide patches as appropriate.

deads2k commented:
@openshift/networking This rebase brought in a new method required for networking plugins.  To make progress, I simply stubbed this in under the `UPSTREAM: openshift/openshift-sdn: <drop>: BREAK OVS PLUGINgit status!` commit.  I don't know the proper fix, can you please provide a patch?

ncdc commented on pkg/build/registry/buildconfiginstantiate/rest.go:
```diff
@@ -209,7 +210,7 @@ func (h *binaryInstantiateHandler) handle(r io.Reader) (runtime.Object, error) {
 		return nil, errors.NewInternalError(fmt.Errorf("unable to connect to server: %v", err))
 	}
-	if err := exec.Stream(r, nil, nil, false); err != nil {
+	if err := exec.Stream([]string{kubeletremotecommand.StreamProtocolV2Name, kubeletremotecommand.StreamProtocolV1Name}, r, nil, nil, false); err != nil {
```
You can use `kubeletremotecommand.SupportedStreamingProtocols`

deads2k replied:
> You can use kubeletremotecommand.SupportedStreamingProtocols

Done.

pweil- commented:
@miminar - you're up to bat on this one since @kargakis and @soltysh are out most of this week.  I'll take a look too.

@legionus @sgallagher - FYI - a rebase in all its glory.  One day the rebase card will have your name on it :smiling_imp:

eparis commented:
@deads2k when is this hoping to land?

sosiouxme commented:
looking at it

danwinship commented:
@deads2k it looks like the ovs plugin Capabilities() method should be:

    return utilsets.NewInt(knetwork.NET_PLUGIN_CAPABILITY_SHAPING)

deads2k commented:
@eparis immediately after 1.2 is tagged before the queue reopens.

dgoodwin commented:
Regarding 9a6a6ed0781e40132deb594bfb2d3481138cb9c4 the upstream commit which changed the interface was https://github.com/kubernetes/kubernetes/commit/79012f6d and it does look like a pretty straightforward rename. 

In a perfect world we should probably remove the "builder" from all the variable names/structs/comments, so I will submit a PR to your branch momentarily @deads2k.

deads2k commented:
> @deads2k it looks like the ovs plugin Capabilities() method should be:
> 
> return utilsets.NewInt(knetwork.NET_PLUGIN_CAPABILITY_SHAPING)

done

deads2k commented:
@jwforres known test-assets problem or have I borked something: https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_integration/56/consoleText ?

stevekuznetsov commented on hack/test-go.sh:
```diff
@@ -186,10 +186,11 @@ else
         \) -name '*_test.go' | xargs -n1 dirname | sort -u | xargs -n1 printf "${OS_GO_PACKAGE}/%s\n")"
 
-        test_packages="${test_packages} ${optional_kubernetes_packages}"
+        test_packages="${optional_kubernetes_packages}"
```
this disables Kube API tests as well

deads2k replied:
> this disables Kube API tests as well

accident, reverting

deads2k replied:
> accident, reverting

Also, how about a "test kube only" mode?

miminar commented on pkg/client/scale.go:
```diff
@@ -35,6 +36,26 @@ func NewDelegatingScaleNamespacer(dcNamespacer DeploymentConfigsNamespacer, sNam
 	}
 }
 
+type delegatingScaleNamespacer2 struct {
```
The higher the number, the bettter the namespacer? Do we really need two?

deads2k replied:
> The higher the number, the bettter the namespacer? Do we really need two?

Casting problems.  We might.  I'll see if I can tweak it.

soltysh replied:
I'll re-rise the issue, I don't see the other being used anywhere else.

deads2k replied:
> I'll re-rise the issue, I don't see the other being used anywhere else.

done

miminar commented on pkg/cmd/server/kubernetes/master_config.go:
```diff
@@ -338,20 +338,24 @@ func BuildKubernetesMasterConfig(options configapi.MasterConfig, requestContextM
 }
 
 // getAPIGroupVersionOverrides builds the overrides in the format expected by master.Config.APIGroupVersionOverrides
```
`s/Config.APIGroupVersionOverrides/Config.APIResourceConfigSource/`

miminar commented on pkg/cmd/server/kubernetes/master_config.go:
```diff
@@ -338,20 +338,24 @@ func BuildKubernetesMasterConfig(options configapi.MasterConfig, requestContextM
 // getAPIGroupVersionOverrides builds the overrides in the format expected by master.Config.APIGroupVersionOverrides
-func getAPIGroupVersionOverrides(options configapi.MasterConfig) map[string]genericapiserver.APIGroupVersionOverride {
-	apiGroupVersionOverrides := map[string]genericapiserver.APIGroupVersionOverride{}
+func getAPIGroupVersionOverrides(options configapi.MasterConfig) genericapiserver.APIResourceConfigSource {
```
Shouldn't this function be renamed as well?

miminar commented on pkg/authorization/authorizer/adapter/attributes.go:
```diff
@@ -70,6 +77,26 @@ func (a AdapterAttributes) GetNamespace() string {
 }
 
+func (a AdapterAttributes) GetName() string {
+	return a.AuthorizationAttributes.GetResourceName()
```
Why [adding `name`](https://github.com/openshift/origin/pull/8484/commits/9cc4e772219d781eba1f6f682a963046a5b0b3b6#diff-d32078125722ccc2c1dddd974cae6216R20) attribute then?

deads2k replied:
> Why adding name attribute then?

removed

miminar commented on pkg/authorization/authorizer/adapter/attributes.go:
```diff
@@ -70,6 +77,26 @@ func (a AdapterAttributes) GetNamespace() string {
+	return tokens[1]
+}
+
+func (a AdapterAttributes) GetResource() string {
```
Godocs.

deads2k replied:
> Godocs.

matches interface, the interface is doc'ed

deads2k commented:
comments addressed and rebased

stevekuznetsov commented:
@deads2k "test only Kube" mode seems useful for you in the rebase but nobody else using `hack/test-go.sh` so I suggest using a `DRY_RUN=true TEST_KUBE=true hack/test-go.sh | sed 'grab the lines you need' | hack/test-go.sh`

deads2k commented:
> @deads2k "test only Kube" mode seems useful for you in the rebase but nobody else using hack/test-go.sh so I suggest using a DRY_RUN=true TEST_KUBE=true hack/test-go.sh | sed 'grab the lines you need' | hack/test-go.sh

You probably meant to say, "useful to anyone doing an upstream pick who ran `test-go.sh` and then realized that he forgot TEST_KUBE=1"....

pweil- commented:
if only we had a makefile that would set this so we didn't have to remember :)

stevekuznetsov commented:
> and then realized that he forgot `TEST_KUBE=1`

I'm not sure *another* mode of operation would help a user if they aren't solid on the current API :)

deads2k commented:
Ha, green!  merge it quick! :)

miminar commented:
LGTM after a very mild and amateurish analysis

liggitt commented on pkg/cmd/server/etcd/etcdserver/server.go:
```diff
@@ -55,11 +55,6 @@ func startEtcd(cfg *config) (<-chan struct{}, error) {
 		return nil, fmt.Errorf("error setting up initial cluster: %v", err)
 	}
 
```
@smarterclayton can you review this part?

smarterclayton replied:
Why was this removed?

smarterclayton replied:
Not a field anymore?

deads2k replied:
> Not a field anymore?

Yeah, you can no longer pass a transport.

liggitt commented on pkg/authorization/authorizer/adapter/attributes.go:
```diff
@@ -70,6 +76,26 @@ func (a AdapterAttributes) GetNamespace() string {
+}
+
+func (a AdapterAttributes) GetResource() string {
+	tokens := strings.SplitN(a.AuthorizationAttributes.GetResource(), "/", 2)
```
this seems like it propagates the bug... `AdapterAttributes` satisfying `oauthorizer.AuthorizationAttributes` with a `GetResource` impl that does not include subresource is scary

we should update `oauthorizer.AuthorizationAttributes` to also have a GetSubresource(), and update `node_auth.go` to set resource and subresource correctly

deads2k replied:
Not now.  Eliminated it satisfying both interfaces.  Perfect round tripping
is still supported on the exact type.  It should never have been anonymous
and it should probably have never tried to satisfy both interfaces.

On Thu, Apr 14, 2016 at 2:19 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/authorization/authorizer/adapter/attributes.go
> <https://github.com/openshift/origin/pull/8484#discussion_r59765891>:
>
> > @@ -70,6 +76,26 @@ func (a AdapterAttributes) GetNamespace() string {
> >  	return a.namespace
> >  }
> >
> > +func (a AdapterAttributes) GetName() string {
> > +	return a.AuthorizationAttributes.GetResourceName()
> > +}
> > +
> > +func (a AdapterAttributes) GetSubresource() string {
> > +	tokens := strings.SplitN(a.AuthorizationAttributes.GetResource(), "/", 2)
> > +	if len(tokens) != 2 {
> > +		return ""
> > +	}
> > +	return tokens[1]
> > +}
> > +
> > +func (a AdapterAttributes) GetResource() string {
> > +	tokens := strings.SplitN(a.AuthorizationAttributes.GetResource(), "/", 2)
>
> this seems like it propagates the bug... AdapterAttributes satisfying
> oauthorizer.AuthorizationAttributes with a GetResource impl that does not
> include subresource is scary
>
> we should update oauthorizer.AuthorizationAttributes to also have a
> GetSubresource(), and update node_auth.go to set resource and subresource
> correctly
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8484/files/c627a3c0e36d1a4023f08e9e6ec3d0c4e62e3421..7e72550dea93a514bf6bfc30e087a46d6a23a3a7#r59765891>
>

deads2k commented:
@liggitt @pweil- I'd liked this merged after we tag and **before** master re-opens.  Let's prioritize reviews if 1.2 is close.

soltysh commented:
I'll reserve tomorrow for a rewiew in that case.

kargakis commented:
> I'll reserve tomorrow for a review in that case.

me too

smarterclayton commented:
We're probably going to strongly control the merge order post fork to
prioritize the key things.

On Wed, Apr 20, 2016 at 11:44 AM, Michail Kargakis <notifications@github.com
> wrote:

> I'll reserve tomorrow for a review in that case.
>
> me too
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8484#issuecomment-212483543>
>

smarterclayton commented on pkg/build/registry/buildconfiginstantiate/rest.go:
```diff
@@ -209,7 +210,7 @@ func (h *binaryInstantiateHandler) handle(r io.Reader) (runtime.Object, error) {
 		return nil, errors.NewInternalError(fmt.Errorf("unable to connect to server: %v", err))
 	}
-	if err := exec.Stream(r, nil, nil, false); err != nil {
+	if err := exec.Stream(kubeletremotecommand.SupportedStreamingProtocols, r, nil, nil, false); err != nil {
```
@ncdc is there a backcompat implication?

ncdc replied:
It's backwards compatible

On Wednesday, April 20, 2016, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/build/registry/buildconfiginstantiate/rest.go
> <https://github.com/openshift/origin/pull/8484#discussion_r60513048>:
>
> > @@ -209,7 +210,7 @@ func (h *binaryInstantiateHandler) handle(r io.Reader) (runtime.Object, error) {
> >  	if err != nil {
> >  		return nil, errors.NewInternalError(fmt.Errorf("unable to connect to server: %v", err))
> >  	}
> > -	if err := exec.Stream(r, nil, nil, false); err != nil {
> > +	if err := exec.Stream(kubeletremotecommand.SupportedStreamingProtocols, r, nil, nil, false); err != nil {
>
> @ncdc <https://github.com/ncdc> is there a backcompat implication?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8484/files/acac24a0bac7bfcf9b7a4adefadb89114d76ecb9..8e14778586be49d76a96c218244278b074fa9766#r60513048>
>

smarterclayton commented on pkg/volume/emptydir/empty_dir_quota.go:
```diff
@@ -8,25 +8,25 @@ import (
 
 var _ volume.VolumePlugin = &EmptyDirQuotaPlugin{}
-var _ volume.Builder = &emptyDirQuotaBuilder{}
+var _ volume.Mounter = &emptyDirQuotaMounter{}
```
@dgoodwin please review this change

dgoodwin replied:
Already did above when deads looped me in. Checked the upstream change, just a simple rename so looks good.

smarterclayton commented on pkg/cmd/util/clientcmd/factory.go:
```diff
@@ -170,7 +170,7 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
 	}
 
-	w.Object = func() (meta.RESTMapper, runtime.ObjectTyper) {
+	w.Object = func(bool) (meta.RESTMapper, runtime.ObjectTyper) {
```
We're going to have to start enabling 3rd party resources soon.

deads2k replied:
> We're going to have to start enabling 3rd party resources soon.

Not during the rebase and preferably not like they've done.  We should make a complete caching discovery client upstream, the right code to use a cache unless it fails on an unknown type.  A discovery `RESTMapper` ought not be hard to build either.

kargakis replied:
follow-up issue?

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/runtime/deep_copy_generator.go:
```diff
@@ -0,0 +1,609 @@
+/*
```
Why?

deads2k replied:
> Why?

At the time?  Time.  The new deep copy generator had multiple bad assumptions about package availability, the old deep copy generator worked, and 1.2 was going to be done two days after I finished this so I wanted reviews, merge, then fix-up.

smarterclayton replied:
Is the follow up issue list linked?

On Wed, Apr 20, 2016 at 10:22 PM, David Eads <notifications@github.com>
wrote:

> In
> Godeps/_workspace/src/k8s.io/kubernetes/pkg/runtime/deep_copy_generator.go
> <https://github.com/openshift/origin/pull/8484#discussion_r60518325>:
>
> > @@ -0,0 +1,609 @@
> > +/*
>
> Why?
>
> At the time? Time. The new deep copy generator had multiple bad
> assumptions about package availability, the old deep copy generator worked,
> and 1.2 was going to be done two days after I finished this so I wanted
> reviews, merge, then fix-up.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8484/files/c1afa8b0bd73fd9704f63c937bb351675987345a..7b2c760d241e2c46d87e4ce977ea65f06644e999#r60518325>
>

deads2k commented:
Image import.  re[test]

deads2k commented:
> We're probably going to strongly control the merge order post fork to
prioritize the key things.

Unless we have concurrent hot fixes, I'd like this next.

smarterclayton commented:
Once this is ready create a PR to smarterclayton:post_12

On Wed, Apr 20, 2016 at 9:41 PM, David Eads <notifications@github.com>
wrote:

> We're probably going to strongly control the merge order post fork to
> prioritize the key things.
>
> Unless we have concurrent hot fixes, I'd like this next.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8484#issuecomment-212689834>
>

kargakis commented on pkg/cmd/cli/cmd/debug.go:
```diff
@@ -199,7 +199,7 @@ func (o *DebugOptions) Complete(cmd *cobra.Command, f *clientcmd.Factory, args [
 		Flatten()
 	if len(o.Filename) > 0 {
-		b.FilenameParam(explicit, o.Filename)
+		b.FilenameParam(explicit, false, o.Filename)
```
Is there an issue to enable these?

soltysh replied:
Looking at the affected files, I doubt we need recursive file visitor for these commands, at least for now.

soltysh commented on hack/test-go.sh:
```diff
@@ -186,6 +186,7 @@ else
             -o -path "${kubernetes_path}/pkg/api/v1"                                                  \
             -o -path "${kubernetes_path}/test/e2e"                                                    \
             -o -path "${kubernetes_path}/cmd/libs/go2idl/client-gen/testoutput/testgroup/unversioned" \
+            -o -path "${kubernetes_path}/pkg/storage/etcd3"                                           \
```
We don't need it right now, since it's not finished, from what I see. But I'd suggest marking this commit as `<drop>` for next rebase.

kargakis replied:
Mind commenting on why are these disabled? Actually, each suite should be followed by such a comment.

deads2k replied:
> Mind commenting on why are these disabled? Actually, each suite should be followed by such a comment.

Flaky upstream and we don't support etcd3.  Comment added and commit is marked drop.

soltysh commented:
I left a few comments, otherwise I'm giving LGTM.

deads2k commented:
Updated the checklist at the top, fixed up, and responded to comments.

@smarterclayton I'd like the two LIGGITT reviews signed off on before squashing down and opening against your branch.

deads2k commented:
well, not to worry.  This has the same image import flake we have in master.

smarterclayton commented:
SDN had to merge for a p0

On Apr 22, 2016, at 7:10 PM, OpenShift Bot <notifications@github.com> wrote:

Origin Action Required: Pull request cannot be automatically merged, please
rebase your branch from latest HEAD and push again

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/8484#issuecomment-213620792>

deads2k commented:
rebased and squashed down

pweil- commented:
LGTM

smarterclayton commented:
After I cut 1.2.0-rc2 this is first in my "i will merge these in order".

smarterclayton commented:
@soltysh pull through is still failing, can you help me track down where
the retry might be needed?  Possibly in the write path of the ServeBlob
code (it might be succeeding once and then failing later)?  Or we may be
exceeding our retries?
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/3221/

On Mon, Apr 25, 2016 at 11:45 AM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test FAILURE (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/3221/)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8484#issuecomment-214406619>
>

deads2k commented:
re[test]

deads2k commented:
re[test]

smarterclayton commented:
[merge] for 1.3.0 and NO, the queue is not open yet for other merges
